pcsx2/ps2/PGPU.cpp

```diff
@@ -0,0 +1,1101 @@
+//Official name is "GFIFO", according to PS1DRV.
+#define PGPU_DAT_FIFO 0x1000F3E0
+//Registers 0x50 - 0x70, 0x90 - 0xB0, 0xD0, 0xF0 are not known to be used for anything.
+/*the following three are from the homebrew PS2SDK, but their names are somewhat misleading:
+#define PGIF_GPU_STAT (0x1000F300)
+#define PGIF_CFIFO_STAT (0x1000F380)
+#define PGIF_CFIFO_DATA (0x1000F3C0)*/
+
+//XXX: These two already exist in IopHw.h, so remove them from here.
+#define	HW_PS1_GPU_DATA 	0x1f801810
+#define	HW_PS1_GPU_STATUS	0x1f801814
+
+
+
+//Bit-fields definitions for the PGPU Status register:
+#define PGPU_STAT_IRQ1	0x01000000
+
+//Specifies the bits in the PGIF Control (0xf380) register that are not writable from EE side, by the PS1DRV.
+#define PGIF_CTRL_RO_MASK 0x00000000
+
+
+#define PGPU_DMA_MADR 0x1F8010A0
+#define PGPU_DMA_BCR  0x1F8010A4
+#define PGPU_DMA_CHCR 0x1F8010A8
+//is there a tadr?
+#define PGPU_DMA_TADR 0x1F8010AC
+
+#define DMA_START_BUSY 0x01000000
+#define DMA_TRIGGER    0x10000000
+//write to peripheral
+#define DMA_DIR_FROM_RAM 0x1
+#define DMA_LINKED_LIST  0x00000400
+#define DMA_LL_END_CODE 0x00FFFFFF
+
+
+
+
+//Local variables for emulation:
+
+static u32 PGpuStatReg = 0; //Read-only to IOP
+//Hold the values of the immediate-response registers.
+u32 pgif1reg = 0;
+u32 pgif2reg = 0;
+u32 pgif3reg = 0;
+u32 pgif4reg = 0;
+
+static u32 PGifCtrlReg = 0;	//PGifIfStat
+static u32 PGifCtrlRegOld = 0;
+static u32 PGpuDataReg = 0; //Read-only to IOP
+
+u32 lastGpuCmd = 0;
+u32 lastGpuData = 0;
+
+//int clrBufCnt = 0;
+
+
+#if 0
+u32 pgpuDmaMadr = 0;
+u32 pgpuDmaBcr = 0;
+u32 pgpuDmaChcr = 0;
+u32 pgpuDmaTadr = 0;
+#else
+#define pgpuDmaMadr HW_DMA2_MADR
+#define pgpuDmaBcr HW_DMA2_BCR
+#define pgpuDmaChcr HW_DMA2_CHCR
+//u32 pgpuDmaChcr = 0;
+#define pgpuDmaTadr HW_DMA2_TADR
+#endif
+
+//Internal flags:
+int PgpuDmaLlAct = 0;
+int PgpuDmaNrActToGpu = 0;
+int PgpuDmaNrActToIop = 0;
+
+u32 pgpuDmaTrAddr = 0;
+u32 wCnt = 0; //
+u32 wordNr = 0; //total number of words
+u32 nextAddr = 0;
+u32 dntPrt = 0;
+
+u32 nrWordsN = 0; //total number of words in Normal DMA
+u32 crWordN = 0; //current word number in Normal DMA
+u32 addrNdma = 0;
+
+
+
+void fillFifoOnDrain(void);
+
+void drainPgpuDmaLl(void);
+void drainPgpuDmaNrToGpu(void);
+void drainPgpuDmaNrToIop(void);
+
+//u32 getUpdPgifCtrlReg(void);
+
+
+
+//Generic FIFO-related:
+struct ringBuf_t {
+	void *buf;
+	int size;
+	int count;
+	int head;
+	int tail;
+};
+
+int ringBufPut(struct ringBuf_t *rb, u32 *data) {
+	if (rb->count < rb->size) { //there is available space
+		*((u32*)rb->buf + rb->head) = *data;
+		if ((++(rb->head)) >= rb->size) rb->head = 0; //wrap back when the end is reached
+		rb->count++;
+	} else {
+		Console.WriteLn("############################# PGIF FIFO overflow! sz= %X", rb->size);
+		return 1; //overflow
+	}
+	return 0;
+}
+
+int ringBufGet(struct ringBuf_t *rb, u32 *data) {
```

Comment:
It's return value doesn't seem to be used, this can be ``void ringBufGet(struct ringBuf_t *rb, u32 *data)`` instead. Same for the above function ``ringBufPut()``